src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+    reference_matrix.resize(_column_segment_begins.size());
+    for (auto& pos_list : reference_matrix) {
+      pos_list.reserve(input_table->row_count());
+    }
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      const auto& chunk = input_table->get_chunk(ChunkID{chunk_id});
+
+      for (size_t segment_id = 0; segment_id < _column_segment_begins.size(); ++segment_id) {
+        const auto column_id = _column_segment_begins[segment_id];
+        const auto column = chunk.get_column(column_id);
+        const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column);
+
+        auto& out_pos_list = reference_matrix[segment_id];
+        auto in_pos_list = ref_column->pos_list();
+        std::copy(in_pos_list->begin(), in_pos_list->end(), std::back_inserter(out_pos_list));
+      }
+    }
+  };
+
+  ReferenceMatrix reference_matrix_left;
+  ReferenceMatrix reference_matrix_right;
+
+  build_reference_matrix(_input_table_left(), reference_matrix_left);
+  build_reference_matrix(_input_table_right(), reference_matrix_right);
+
+  /**
+   * Init the virtual pos lists
+   */
+  VirtualPosList virtual_pos_list_left(_input_table_left()->row_count(), 0u);
+  std::iota(virtual_pos_list_left.begin(), virtual_pos_list_left.end(), 0u);
+  VirtualPosList virtual_pos_list_right(_input_table_right()->row_count(), 0u);
+  std::iota(virtual_pos_list_right.begin(), virtual_pos_list_right.end(), 0u);
+
+  /**
+   * Sort the virtual pos lists, so we can merge them
+   */
+  std::sort(virtual_pos_list_left.begin(), virtual_pos_list_left.end(),
+            VirtualPosListCmpContext{reference_matrix_left});
+  std::sort(virtual_pos_list_right.begin(), virtual_pos_list_right.end(),
+            VirtualPosListCmpContext{reference_matrix_right});
+
+  /**
+   * Build result table
+   */
+  auto left_idx = size_t{0};
+  auto right_idx = size_t{0};
+  const auto num_left = virtual_pos_list_left.size();
+  const auto num_right = virtual_pos_list_right.size();
+
+  // Somewhat random way to decide on a chunk size.
+  const auto out_chunk_size = std::max(_input_table_left()->chunk_size(), _input_table_right()->chunk_size());
+
+  auto out_table = Table::create_with_layout_from(_input_table_left(), out_chunk_size);
+
+  std::vector<std::shared_ptr<PosList>> pos_lists(reference_matrix_left.size());
+  std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+
+  // Adds one row to the pos_lists currently being filled
+  const auto emit_row = [&](const ReferenceMatrix& reference_matrix, size_t row_idx) {
+    for (size_t pos_list_idx = 0; pos_list_idx < pos_lists.size(); ++pos_list_idx) {
+      pos_lists[pos_list_idx]->emplace_back(reference_matrix[pos_list_idx][row_idx]);
+    }
+  };
+
+  // Turn 'pos_lists' into a new chunk and append it to the table
+  const auto emit_chunk = [&]() {
```

Comment:
I find that the inline lambdas make it difficult for me to follow the structure of the code. At this point, a lambda is defined that is used much later. Here, I read something about how it adds all entries from the pos lists, but so far I have not seen any "unique" stuff.

Reply:
> If a private/lambda is long, and appears incongruous with the code calling it, this can distract the reader from a greater pattern in the code. Imagine if you were reading one of these StackOverflow answers and suddenly had to stop to read StackEnglish's description of null and void to make sense of the answer. You'd really rather just see a hyperlink of "optional information" than a copy/paste of the contents.

https://stackoverflow.com/a/18805673/2204581

Reply:
you're right, absolutely overdid it with the lambdas here, will refactor